### PR TITLE
Classrooms - adding API fields

### LINKED ISSUE #6032: Classrooms - Fields populated by API
Split off from https://github.com/uiowa/uiowa/issues/5769

# Proposed solution

Add the following fields, which will be populated via API.

- [x] **Classroom Type - single value
    - Used as filter
    - Data comes from roomType in https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraRoomData/Abw/240
    - https://classrooms.uiowa.edu/admin/structure/taxonomy/classroom_type
    - we should link these room types to https://classrooms.prod.drupal.uiowa.edu/types-university-classrooms
- [x] **Max Occupancy - this should only contain a number
    - Used as a filter
    - comes from maxOccupancy https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraRoomData/Abw/240
- [x] **Features
    - They would like to break this into two groups - Room Features and Technology features
    - Currently represented as one filter in mockup but two separate would be ideal
        - https://classrooms.prod.drupal.uiowa.edu/home/find-university-classroom#feature
    - MAUI just contains all of these in one data set, not differentiated between room and technology
    - Not linking to these, rabbit hole if using a reference?
    - I've asked them to provide us a list of all of the room features and all of the tech features so we have a way to determine the mappings
    - There will be a couple features in MAUI they we will want to avoid displaying for security/safety purposes. I still have a question out related to this. Ideally, they have control of the list of tech features and room features on the website (taxonomy, etc.) so that if a new feature is added to MAUI they can add that to the site so that the mapping will work for the new feature. Same if they decide to remove a feature. They could delete the item from the list (taxonomy term, etc) and then that feature would no longer be mapped or show on the site.
- [x] **Responsible Unit - List of Department/Units
    - Fields 
        - Name
        - Description - not sure if used currently - they would like to keep but not output if empty
        - Scheduler - text
        - Email
        - Phone
        - Link - title and URL
    - Responsible unit comes from MAUI acadOrgUnitName. Additional unit information will be manually managed by Classrooms folks - can this additional unit info be brought over in the content migration?
- [x] **Room Name - text field - i believe this is if there is an alternative name for a room (e.g. Faculty Studio https://registrar.uiowa.edu/vox-5468) - Yes and will need to be included on room display - w10 PBB example.
- [x] **Instructional Room Category? - This can be the field used internally to designate University Classroom vs Programmed Classroom.
- [ ] Room schedule - from Astra - may not need any structure for this but will need displayed? If no structure needed this may make sense to be in a future slice.

### LINKED ISSUE #6041: Classrooms - Remaining Fields
# Additional Info

- Tile Details
    - multiple items - text field
    - when we migrate we will need to combine the existing two fields into one
    - There does not need to be a filter on the search page for this
- Design Details
    - multiple items - text field
    - when we migrate we will need to combine the existing two fields into one
    - There does not need to be a filter on the search page for this
- Classroom Furniture
    - multiple items - text field
    - when we migrate we will need to combine the existing two fields into one
    - There does not need to be a filter on the search page for this
- Scheduling Region - comes from MAUI room data API https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraRoomData/ajb/e105
    - "SC SW" is an example from the room in the API link above
    - From the regionList
    - This sounds to me like a taxonomy similar to the features where they will need to populate the list of terms, have control over how it outputs on the page vs how it is in the API (API Name field).
    - There will be a filter on the search page to filter by scheduling region.
    - https://classrooms.prod.drupal.uiowa.edu/university-classroom-scheduling-regions

# Proposed solution

### PR DESCRIPTION
**Pull request by e-marie-w**
Resolves #6032
Resolves https://github.com/uiowa/uiowa/issues/6041

<details><summary>Todo</summary>

- [x] Setup rabbit hole/rabbit_hole taxonomy for the vocabularies we don't want visitors to see (access denied) (all but units)
- [x] Remove "link entity label" for room field renders (all but units)
- [x] Make sure webmasters have add/edit/delete for room_features, technology_features, units (not room types)
- [x] Decide what to do about room schedule https://github.com/uiowa/uiowa/issues/6032#issuecomment-1358441441
- [x] Add help text to disabled fields and related taxonomies
 </details>

# How to test

- Check out classrooms_api_fields and pull
- `ddev blt ds --site=classrooms.uiowa.edu`
- `ddev drush @classrooms.local uli`
- Create a room - select a building, add a room ID, and save (can get rooms from https://classrooms.uiowa.edu/)
- Several fields should now have values. Go back to Edit and open Imported fieldgroup to see what fields did/didn't pull data.
- Confirm saved field data links up to the API - https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraRoomData/{building_id}/{room_id}
- For the 5 taxonomy fields (Responsible Unit, Room Features, Room Type, Scheduling Region, Technology Features) you'll need to add some of the terms from the API data to the taxonomies before saving again to see the data imported. 
- Masquerade as different roles, only webmasters should be able to view/add/delete taxonomy terms. 
- Also check rabbit hole is set up correctly. - trying viewing an added taxonomy term for anything other than the units - for example https://classrooms.uiowa.ddev.site/room-types/auditorium - in an incognitio window. Page should return access denied. Can also test that a units taxonomy term is accessible in incognito.

### CODE REVIEW

File: docroot/sites/classrooms.uiowa.edu/modules/classrooms_core/classrooms_core.module [hunk 1]
```diff
@@ -44,4 +68,32 @@ function classrooms_core_entity_presave(EntityInterface $entity): void {
...
+      if (strlen($data[0]->roomName) > 1) {
+        $entity->set('field_room_name', $data[0]->roomName);
+      }
+    }
+
+    // Select list field not working yet.
+    //    if ($entity->hasField('field_room_instruction_category') && isset($data[0]->roomCategory)) {
+    //      $field_definition = $entity->getFieldDefinition('field_room_instruction_category')->getFieldStorageDefinition();
+    //      $field_allowed_options = options_allowed_values($field_definition, $entity);
+    //      if (in_array($data[0]->roomCategory, $field_allowed_options)) {
```
Comment (joewhitsitt) [inline docroot/sites/classrooms.uiowa.edu/modules/classrooms_core/classrooms_core.module]:
```suggestion
    //      if (array_key_exists($data[0]->roomCategory, $field_allowed_options)) {
```

couldn't leave without understanding why this wouldn't work. we want to compare the key, not the array value.

File: config/sites/classrooms.uiowa.edu/config_split.patch.user.role.webmaster.yml [hunk 2]
```diff
@@ -2,18 +2,34 @@ adding:
...
     - 'edit own room content'
+    - 'edit terms in room_features'
+    - 'edit terms in technology_features'
+    - 'edit terms in units'
     - 'enter room revision log entry'
     - 'override room authored by option'
     - 'override room authored on option'
     - 'override room published option'
     - 'override room sticky option'
+    - 'rabbit hole administer taxonomy_term'
```
Comment (joewhitsitt) [inline config/sites/classrooms.uiowa.edu/config_split.patch.user.role.webmaster.yml]:
```suggestion
```

we don't want webmasters to configure rabbit hole. just want them to bypass any blocked/redirected paths associated with RH

**Comment by bspeare** [comment]
Pulling to test.

**Comment by bspeare** [comment]
@e-marie-w can you advise on this one? What is the "API Mapping" field supposed to be populated with?

> For the 5 taxonomy fields (Responsible Unit, Room Features, Room Type, Scheduling Region, Technology Features) you'll need to add some of the terms from the API data to the taxonomies before saving again to see the data imported.

Screen Shot 2022-12-21 at 3 50 09 PM

**Comment by bspeare** [comment]
Seeing "Room types" with different permissions for webmasters, is that correct?

Screen Shot 2022-12-21 at 3 56 11 PM

**Comment by joewhitsitt** [comment]
For those taxonomies, we are pulling from one data list and splitting out into different buckets. If you look at this one below, you can grab some of the strings from featureList and put into any of those taxonomies to have it get mapped. The customer is familiar with these but we will provide some handover training so that they can have "Air conditioning" as the human label vs "A/C"

https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraRoomData/Abw/240

>Seeing "Room types" with different permissions for webmasters, is that correct?

Features is supposed to be managed (add/edit/delete) by webmasters. Room types should be a fairly static list for now and shouldn't need their management.

**Comment by bspeare** [comment]
@joewhitsitt thanks for explaining.  The testing all works as you have described.  Maybe we should add some help text on that field for future users `/admin/structure/taxonomy/manage/room_features/overview/fields/taxonomy_term.room_features.field_api_mapping` like "Enter the MAUI featureList name here so that..." or something.

**Comment by e-marie-w** [comment]
> Maybe we should add some help text on that field for future users

Great idea! I can get something added after standup.

**Comment by e-marie-w** [comment]
Wait, since the help text can be specified for each instance of the reused API Mapping field, I could just set the necessary mapping as the default value for each taxonomy, right?

**Comment by joewhitsitt** [comment]
> Wait, since the help text can be specified for each instance of the reused API Mapping field, I could just set the necessary mapping as the default value for each taxonomy, right?

Each taxonomy here is pulling from the same feed (registrar/courses/AstraRoomData). Some are pulling from different properties regionList vs featureList, but the actual values for the API Mapping field will be different for every term ("A/C", "Computer - MAC", "West") so I don't think you would want a default value. Your help text can differ per instance and be slightly different based on the feed property they should be looking at.

**Comment by e-marie-w** [comment]
> > Wait, since the help text can be specified for each instance of the reused API Mapping field, I could just set the necessary mapping as the default value for each taxonomy, right?
> 
> Each taxonomy here is pulling from the same feed (registrar/courses/AstraRoomData). Some are pulling from different properties regionList vs featureList, but the actual values for the API Mapping field will be different for every term ("A/C", "Computer - MAC", "West") so I don't think you would want a default value. Your help text can differ per instance and be slightly different based on the feed property they should be looking at.

Crap I forgot that's how it worked okay ignore last commit going to fix that real quick.

File: config/sites/classrooms.uiowa.edu/field.field.taxonomy_term.room_features.field_api_mapping.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+dependencies:
+  config:
+    - field.storage.taxonomy_term.field_api_mapping
+    - taxonomy.vocabulary.room_features
+id: taxonomy_term.room_features.field_api_mapping
+field_name: field_api_mapping
+entity_type: taxonomy_term
+bundle: room_features
+label: 'API Mapping'
+description: 'Enter the MAUI featureList name of the feature to add, for example A/C.'
```
Comment (joewhitsitt) [inline config/sites/classrooms.uiowa.edu/field.field.taxonomy_term.room_features.field_api_mapping.yml]:
```suggestion
description: 'Enter value from the MAUI <em>featureList</em> property to map this term to (e.g. <em>A/C</em>).
```

File: config/sites/classrooms.uiowa.edu/field.field.taxonomy_term.scheduling_regions.field_api_mapping.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+dependencies:
+  config:
+    - field.storage.taxonomy_term.field_api_mapping
+    - taxonomy.vocabulary.scheduling_regions
+id: taxonomy_term.scheduling_regions.field_api_mapping
+field_name: field_api_mapping
+entity_type: taxonomy_term
+bundle: scheduling_regions
+label: 'API Mapping'
+description: 'Enter the MAUI regionList name of the region to add, for example West or Pentacrest.'
```
Comment (joewhitsitt) [inline config/sites/classrooms.uiowa.edu/field.field.taxonomy_term.scheduling_regions.field_api_mapping.yml]:
```suggestion
description: 'Enter value from the MAUI <em>regionList</em> property to map this term to (e.g. <em>Pentacrest</em>).'
```

File: config/sites/classrooms.uiowa.edu/field.field.taxonomy_term.technology_features.field_api_mapping.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+dependencies:
+  config:
+    - field.storage.taxonomy_term.field_api_mapping
+    - taxonomy.vocabulary.technology_features
+id: taxonomy_term.technology_features.field_api_mapping
+field_name: field_api_mapping
+entity_type: taxonomy_term
+bundle: technology_features
+label: 'API Mapping'
+description: 'Enter the MAUI featureList name of the feature to add, for example Computer - Windows.'
```
Comment (joewhitsitt) [inline config/sites/classrooms.uiowa.edu/field.field.taxonomy_term.technology_features.field_api_mapping.yml]:
```suggestion
description: 'Enter value from the MAUI <em>featureList</em> property to map this term to (e.g. <em>Computer - MAC</em>).'
```
This is a good example to emphasize the case sensitivity of the mapping.

**Review by bspeare — APPROVED** [review]
Tested this out and it works great!  Thanks for adding the extra help text.

**Comment by e-marie-w** [comment]
Thanks for testing @bspeare!

**Pull request merged**